### ISSUE TITLE
Improve GUID matching

### ISSUE THREAD

**Issue by offalynne** [offalynne/Input#53]
I know we've discussed this, wanting to consolidate thoughts on it in one place in case it's of interest.

Maps should be matched based on full GUID string and not just VID PID pair. The default GUID format shouldn't be assumed when extracting VID & PID.

The exception is on Windows, where VID & PID is adequate since GM's using the older format that omits other info. For type identification it's adequate (assuming the default GUID format was consumed properly), since that's all SDL's data uses. 

Windows (old format):
```
********000000000000504944564944`
^   ^
VID PID
```

Non-Windows platforms GUIDs can take one of three formats:

Default:
```
****0000****0000****0000****0000
^       ^       ^       ^
Driver  Vendor  Product Revision
```

In this format all fields are important, for example a Bluetooth mode device (driver `0500`) may map different in USB mode (driver `0300`), and revision is heavily relied on for differentiating otherwise identically-reporting devices.

Fallback (Mac & Linux):
```
****0000************************
^       ^
Driver  Devce name in UTF-8, right-padded
```

Fallback (Android):
```
********************************
^
Device name CRC as integer string
```

The most straightforward way to check for default format seems to me (at a glance) to check that the full GUID is not zeroed, but the expected padding after all 4 fields is.

If a full GUID comparison fails and VID & PID are found you can (but don't need to) fallback on VID & PID alone to find "nearest match".

Currently (iirc) you've got a platform-agnostic fallback if no same-platform match is made in the external or inbuilt db. The only (relatively) safe mapping fallback between platforms is between Windows and Linux.

**Comment by JujuAdams** [comment]
Marking as bug as _technically_ we're doing things wrong atm and "bug" will scare me into fixing it sooner

**Comment by JujuAdams** [comment]
Implemented exact GUID matching as the primary database check today. Safer VID+PID extraction will come a little later.

> Currently (iirc) you've got a platform-agnostic fallback if no same-platform match is made in the external or inbuilt db. The only (relatively) safe mapping fallback between platforms is between Windows and Linux.

Shall I just remove that? I have no love for this feature.

**Comment by offalynne** [comment]
Sounds good

**Comment by JujuAdams** [comment]
Removed the OS-agnostic fallback, which also allows us to cut the memory usage substantially!

In other news, it seems like the SDL2 database has some issues where a few of the GUIDs in the database don't match the pattern proposed. I wonder if there's been some human error in inserting these GUIDs.

For example `30000004c050000e60c0000000100000` is a "Sony DualSense" controller on MacOS, and it looks like a zero has teleported from the start of the GUID to the end. `03000000c82d00015900000000000000` and `03000000c82d00065280000000000000` also look like their PID has been offset in the middle of the string. There are a few others as well.

**Comment by offalynne** [comment]
Super weird. Gonna investigate where I’m able. Even though it’s “wrong” if it’s wrong consistently that’s still useful in it’s own way ofc.

**Comment by offalynne** [comment]
Fixed :stuck_out_tongue_closed_eyes:

Feel free to point out any others

**Comment by JujuAdams** [comment]
Couldn't see others at this time

Improved GUID matching will be available in v3.2.3
